core, api, benchmarks: Random acts of garbage reduction

njhill opened the pull request:
I noticed some opportunities to reduce allocations on hot paths

ejona86 commented on core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:
```diff
@@ -64,92 +72,109 @@ public int readableBytes() {
-  public CompositeReadableBuffer readBytes(int length) {
+  public ReadableBuffer readBytes(int length) {
+    if (length <= 0) {
+      return ReadableBuffers.empty();
```
It's unclear if this change improves or worsens performance. Now the type of the return value changes, so the JIT will not optimize the calling code as much. On the other hand, there is less indirection when `length` is present in a single buffer.

ejona86 commented on core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:
```diff
@@ -189,6 +211,16 @@ private void execute(ReadOperation op, int length) {
+    try {
+      return execute(op, length, dest, value);
+    } catch (IOException e) {
+      throw new RuntimeException(e); // shouldn't happen
```
We use AssertionError for impossible internally-guaranteed conditions.

ejona86 commented on core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:
```diff
@@ -202,32 +234,11 @@ private void advanceBufferIfNecessary() {
-    }
-
-    abstract int readInternal(ReadableBuffer buffer, int length) throws IOException;
+  private interface ReadOperation<T> {
```
Don't know if we care, but we could have a `NonThrowingReadOperation` that extends this interface and doesn't have `throws IOException`. `executeNoThrow()` could then have that as a parameter. That would preserve the checked exception checking.

njhill replied:
I may not completely follow, but I don't see that this would buy much. Unless we wanted to also duplicate the `execute()` method contents which wouldn't be ideal.

ejona86 replied:
No need to duplicate execute(). `execute()` and `executeNoThrow()` would work as you have it here. The difference is that `executeNoThrow()` could only be called with a ReadOperation that actually doesn't throw IOExceptions.

ejona86 commented on core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:
```diff
@@ -64,92 +72,109 @@ public int readableBytes() {
+  };
+
   @Override
   public void readBytes(final byte[] dest, final int destOffset, int length) {
```
You might want to clean up the now-unnecessary `final`s. Not a big deal.

ejona86 commented on core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:
```diff
@@ -202,32 +234,11 @@ private void advanceBufferIfNecessary() {
-
-    abstract int readInternal(ReadableBuffer buffer, int length) throws IOException;
+  private interface ReadOperation<T> {
+    int read(ReadableBuffer buffer, int length, T dest, int value) throws IOException;
```
Somewhere the relationship between `value` and the returned `int` needs to be mentioned, since that might be surprising and doesn't "jump out" obviously in the code.

ejona86 reviewed (commented):
The UnsafeByteOperations change is the only reason I'm not approving right now (although I wouldn't merge). I do think there are some tweaks that could be made.

ejona86 commented on core/src/main/java/io/grpc/internal/MessageFramer.java:
```diff
@@ -218,28 +218,26 @@ private int writeKnownLengthUncompressed(InputStream message, int messageLength)
               String.format("message too large %d > %d", messageLength , maxOutboundMessageSize))
           .asRuntimeException();
     }
-    ByteBuffer header = ByteBuffer.wrap(headerScratch);
```
FWIW, this allocation should exist on the stack.

voidzcy replied:
> FWIW, this allocation should exist on the stack.

What do you mean by "this allocation"?

ejona86 replied:
ByteBuffer.wrap() creates a new object.

voidzcy commented on core/src/main/java/io/grpc/inprocess/InProcessTransport.java:
```diff
@@ -96,7 +97,8 @@
   @GuardedBy("this")
-  private Set<InProcessStream> streams = new HashSet<>();
+  private Set<InProcessStream> streams = Collections.newSetFromMap(
+          new IdentityHashMap<InProcessStream, Boolean>());
```
Can do the same for [`CronetClientTransport`](https://github.com/grpc/grpc-java/blob/d52b359631ee1ac46f47d72d1015e1dd6ce9312c/cronet/src/main/java/io/grpc/cronet/CronetClientTransport.java#L56-L57).

njhill commented:
Thanks @ejona86! I will address the comments when I get a chance but likely won't be until next week.

ejona86 commented:
@njhill, that's more than fine. I'm sorry it took so very long for us to review.

And in case it wasn't clear, I'm fine with accepting the garbage reduction changes that probably don't reduce garbage. It's a PITA to check whether the JIT is actually doing what we hope and that still depends on what side of the bed the JIT woke up on. Since the changes don't harm readability, they seem fine. But I pointed them out since the value of those sorts of changes is questionable and we may not want to get out-of-hand with that class of change.

njhill commented on api/src/main/java/io/grpc/CallOptions.java:
```diff
@@ -61,7 +61,7 @@
   private String compressorName;
 
-  private Object[][] customOptions = new Object[0][2];
+  private Object[][] customOptions;
```
It's already the case that there will be a single copy since the constructors are private and all instances derive from `DEFAULT`.

njhill commented on benchmarks/src/main/java/io/grpc/benchmarks/Utils.java:
```diff
@@ -252,7 +253,7 @@ public static SimpleResponse makeResponse(SimpleRequest request) {
       }
 
-      ByteString body = ByteString.copyFrom(new byte[request.getResponseSize()]);
+      ByteString body = UnsafeByteOperations.unsafeWrap(new byte[request.getResponseSize()]);
```
Sure I can revert it. I just thought it might help to isolate the actual grpc parts within the benchmarks, assuming that they're intended more for relative measurements than simulating a particular real-world use case.

njhill commented on core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:
```diff
@@ -50,9 +58,9 @@ public void addBuffer(ReadableBuffer buffer) {
-    while (!compositeBuffer.buffers.isEmpty()) {
-      ReadableBuffer subBuffer = compositeBuffer.buffers.remove();
-      buffers.add(subBuffer);
+    Queue<ReadableBuffer> otherBuffers = compositeBuffer.buffers;
```
I agree it's unlikely to make much/any difference especially here, but seems cleaner and good general practice. I.e. in theory less pointer chasing in the loop. Or could be my OCD.

ejona86 replied:
I don't think I agree with "good general practice." I expect the optimizer to do loop invariant hoisting. For simple cases like this I'd put it in the same category as 'optimizing' `/2` with `>>1` or simplifying `1 + 2` to `3`. But this isn't even that hot of a loop. With reasonably-sized buffers, it seems impossible to ever notice this optimization, even if we assume the processor isn't faster executing the later occurrences or the interpreter was being used.

The "cleaner" part for me was combining `buffers.add()` and the `remove()` onto the same line. But the original author could have done that and actively chose not to. But adding `otherBuffers` actually the change overall less clean to my eyes. In any case the original code seems reasonable as-is, so it seems best not to touch it.

njhill replied:
Sure, I'll revert it! Yes the "cleanliness" is probably subjective, I have a general preference for minimizing indirections, other things being equal (regardless of perf implications). I don't quite agree that it's in the same category as `>>1`, `1 + 2` -> `3` since those have a more obvious readability tradeoff.

njhill commented:
@ejona86 I think I've addressed all the comments, PTAL

I did also try to rebase this branch but it wouldn't let me push that:
```
!	refs/heads/garbage:refs/heads/garbage	[remote rejected] (refusing to allow an OAuth App to create or update workflow `.github/workflows/gradle-wrapper-validation.yml` without `workflow` scope
```

ejona86 commented:
> I did also try to rebase this branch but it wouldn't let me push that:

@njhill, that error doesn't look related to the force push. We have a github workflow now, so it looks like that error would apply every time you try to push to your repo with newer PRs. I don't know how you are authenticating, but maybe you need to go through that process again so it can request more more scopes.

ejona86 commented:
@voidzcy, could you review?

voidzcy commented:
CI failures need to be addressed (missing imports).

ejona86 commented:
Thank you @njhill!
